benchmark tests panic with 96 bytes of unexpected leftover memory

Seen in `sql.test.stderr` from [this run](https://teamcity.cockroachdb.com/viewLog.html?buildTypeId=Cockroach_BenchmarkTests&buildId=37818) of the benchmark tests.

```
161028 07:43:02.977208 1986260 sql/mon/mem_usage.go:86  [n1,client=127.0.0.1:52114] memory usage increases to 64 KiB (+188)
I161028 07:43:02.977346 1986260 sql/mon/mem_usage.go:86  [n1,client=127.0.0.1:52114] memory usage increases to 128 KiB (+188)
I161028 07:43:02.977613 1986260 sql/mon/mem_usage.go:86  [n1,client=127.0.0.1:52114] memory usage increases to 256 KiB (+188)
I161028 07:43:02.978144 1986260 sql/mon/mem_usage.go:86  [n1,client=127.0.0.1:52114] memory usage increases to 512 KiB (+188)
panic: unexpected leftover memory: 96 bytes [recovered]
    panic: unexpected leftover memory: 96 bytes

goroutine 1986045 [running]:
panic(0x17082a0, 0xc421a169f0)
    /usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc42e6e8000)
    /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:185 +0x6e
panic(0x17082a0, 0xc421a169f0)
    /usr/local/go/src/runtime/panic.go:458 +0x243
github.com/cockroachdb/cockroach/pkg/sql/mon.(*MemoryUsageMonitor).StopMonitor(0xc42023b550, 0x7f06f5bc1380, 0xc43e783ac0)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/mon/mem_usage.go:130 +0x2a2
github.com/cockroachdb/cockroach/pkg/sql.(*Session).Finish(0xc42023ad00, 0xc4222ad540)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/session.go:193 +0x99
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).finish(0xc4267a4000, 0x7f06f5bc1340, 0xc42e642510)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:154 +0x7e
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn(0xc42957a5a0, 0x7f06f5bc1340, 0xc42e642510, 0x23efda0, 0xc427bec380, 0x0, 0x0)
    /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:245 +0x659
github.com/cockroachdb/cockroach/pkg/server.(*Server).Start.func8.1(0x23efe00, 0xc42e87e000)
    /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:488 +0x106
github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1(0xc42e6e8000, 0xc4201741a0, 0x23efe00, 0xc42e87e000, 0xc42815a020)
    /go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:136 +0x95
created by github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith
    /go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:138 +0x277
```

--- a/pkg/sql/executor.go
+++ b/pkg/sql/executor.go
@@ -107,7 +107,12 @@ type StatementResults struct {
 
 // Close ensures that the resources claimed by the results are released.
 func (s *StatementResults) Close() {
-	for _, r := range s.ResultList {
+	s.ResultList.Close()
+}
+
+// Close ensures that the resources claimed by the results are released.
+func (rl ResultList) Close() {
+	for _, r := range rl {
 		r.Close()
 	}
 }
@@ -564,6 +569,10 @@ func (e *Executor) execRequest(session *Session, sql string, copymsg copyMsg) St
 			}
 
 			var err error
+			if results != nil {
+				// Some results were produced by a previous attempt. Discard them.
+				ResultList(results).Close()
+			}
 			results, remainingStmts, err = runTxnAttempt(e, planMaker, origState, txnState, opt, stmtsToExec)
 
 			// TODO(andrei): Until #7881 fixed.

--- a/pkg/sql/txn_restart_test.go
+++ b/pkg/sql/txn_restart_test.go
@@ -505,9 +505,21 @@ CREATE TABLE t.test (k INT PRIMARY KEY, v TEXT, t DECIMAL);
 	// of txns (statements batched together with the BEGIN stmt) - are retried.
 	// We also exercise the SQL cluster logical timestamp in here, because
 	// this must be properly propagated across retries.
+	//
+	// The SELECT within the transaction also checks that discarded
+	// intermediate result sets are properly released: the result set it
+	// produces is accounted for by the session monitor, and if it is
+	// not properly released upon a retry the monitor will cause the
+	// server to panic (and thus the test to fail) when the connection
+	// is closed.
+	//
+	// TODO(knz) This test can be made more robust by exposing the
+	// current allocation count in monitor and checking that it has the
+	// same value at the beginning of each retry.
 	if _, err := sqlDB.Exec(`
 INSERT INTO t.test(k, v, t) VALUES (1, 'boulanger', cluster_logical_timestamp());
 BEGIN;
+SELECT * FROM t.test;
 INSERT INTO t.test(k, v, t) VALUES (2, 'dromedary', cluster_logical_timestamp());
 INSERT INTO t.test(k, v, t) VALUES (3, 'fajita', cluster_logical_timestamp());
 END;

